sdcm/utils/remote_logger.py

```diff
@@ -302,6 +315,134 @@ def _logger_cmd(self) -> str:
+                timestamps=True,
+                follow=True,
+                since_seconds=since_seconds,
+                # NOTE: need to set a timeout, because GKE's 'pod_log' API tends to hang
+                _request_timeout=self.READ_REQUEST_TIMEOUT,
+                _preload_content=False
+            )
+            self._log_reader = self._read_log_lines(stream)
+            self._reread_logs_till_last_logged_timestamp()
+        except (k8s.client.rest.ApiException, StopIteration) as exc:
+            self._log.warning(
+                "'_open_stream()': failed to open pod log stream:\n%s", exc)
+            # NOTE: following is workaround for the error 401 which may happen due to
+            #       some config data corruption during the forced socket connection failure
+            self._k8s_core_v1_api = KubernetesOps.core_v1_api(self._parent_cluster.k8s_cluster.get_api_client())
+            raise ConnectionError(str(exc)) from None
+
+    def _reread_logs_till_last_logged_timestamp(self):
+        """Upon reconnection, reread the log till last logged timestamp is found to avoid duplication"""
+        if not self._last_log_timestamp:
+            return
+        log_timestamp, line = next(self._log_reader)
+        if isoparse(log_timestamp) >= isoparse(self._last_log_timestamp):
+            self._log.debug("Current timestap is greater than last logged timestamp. No need to reread logs")
+            self._write_log_line(log_timestamp, line)
+            return
+        self._log.debug("Rereading logs till %s", self._last_log_timestamp)
+        log_timestamp = "not read yet"
+        while True:
+            for _ in range(1000):
+                log_timestamp = next(self._log_reader)[0]
+                if log_timestamp == self._last_log_timestamp:
+                    return
+            # each 1000 lines, check if we didn't pass the date, this should not happen and possibly we can drop this code after some time
+            if isoparse(log_timestamp) >= isoparse(self._last_log_timestamp):
+                TestFrameworkEvent(source="K8sClientLogger",
+                                   message=f"Failed to find last log timestamp in the log stream for {self._pod_name} after reconnection."
+                                           f" Possibly some log lines are missed or duplicated. Last reread timestamp: {log_timestamp}",
+                                   severity=Severity.WARNING).publish()
+                return
+
+    def _write_log_line(self, timestamp, line):
+        self._last_read_timestamp = time.time()
+        self._last_log_timestamp = timestamp
+        self._file_object.write(line)
+
+    def _read_log_lines(self, stream) -> Generator[tuple[str, str], None, None]:
+        """Reads log in chunks and split to lines. Returns a tuple of timestamp and log line"""
+        buffer = ''
+        while True:
+            chunk = stream.read(self.CHUNK_SIZE).decode("utf-8")
```

Comment:
I don't think is guaranteed, as in case of closing stream lines would be generated anyway. And in case of reconnection, we would re-read them again. The real problem is that we could wait long for these lines to get to SCT.

I refactored this to use more native solution, see added commit and let me know what you think.

Reply:
Looks ok, need to see it's work in a CI job.

Reply:
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/multitenant-eks/12/

Reply:
for me, from logs perspective this job looks fine

Reply:
Logs are not complete.

Streamed logs for `loaders-2-0` last command are following:

```
...
java.io.IOException: Operation x10 on key(s) [374e38324d334f4c3030]: Error executing: (NoHostAvailableException): All host(s) tried for query failed (no host was tried)

	at org.apache.cassandra.stress.Operation.error(Operation.java:141)com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (no host was tried)

	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:119)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:473)
java.io.IOException: Operation x10 on key(s) [3835304d364c37313530]: Error executing: (NoHostAvailableException): All host(s) tried for query failed (no host was tried)

	at org.apache.cassandra.stress.Operation.error(Operation.java:141)
	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:119)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.ja
```

Pod `namespaces/sct-loaders-2/pods/sct-loaders-2-0-pod-2` has following logs ending:

```
...
java.io.IOException: Operation x10 on key(s) [374e38324d334f4c3030]: Error executing: (NoHostAvailableException): All host(s) tried for query failed (no host was tried)

	at org.apache.cassandra.stress.Operation.error(Operation.java:141)com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (no host was tried)

	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:119)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:473)
java.io.IOException: Operation x10 on key(s) [3835304d364c37313530]: Error executing: (NoHostAvailableException): All host(s) tried for query failed (no host was tried)

	at org.apache.cassandra.stress.Operation.error(Operation.java:141)
	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:119)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:101)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:109)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:264)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:473)
ERROR 19:10:01,325 [Control connection] Cannot connect to any host, scheduling retry in 2000 milliseconds
WARN  19:10:01,390 Not using advanced port-based shard awareness with adc37a2d24a9e407c857ef164daa059e-1115191262.eu-north-1.elb.amazonaws.com:9142:de527fd7-8303-4d49-b3bb-e916c019d1c8.cql.sct-cluster-2.sct.scylladb.com because it's disabled in configuration
WARN  19:10:01,396 Not using advanced port-based shard awareness with adc37a2d24a9e407c857ef164daa059e-1115191262.eu-north-1.elb.amazonaws.com:9142:7f7c14e0-eb1c-450a-8502-703c0d0acd3d.cql.sct-cluster-2.sct.scylladb.com because it's disabled in configuration
```

So, here is the missing part:

```
                                                                                             va:264)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:473)
ERROR 19:10:01,325 [Control connection] Cannot connect to any host, scheduling retry in 2000 milliseconds
WARN  19:10:01,390 Not using advanced port-based shard awareness with adc37a2d24a9e407c857ef164daa059e-1115191262.eu-north-1.elb.amazonaws.com:9142:de527fd7-8303-4d49-b3bb-e916c019d1c8.cql.sct-cluster-2.sct.scylladb.com because it's disabled in configuration
WARN  19:10:01,396 Not using advanced port-based shard awareness with adc37a2d24a9e407c857ef164daa059e-1115191262.eu-north-1.elb.amazonaws.com:9142:7f7c14e0-eb1c-450a-8502-703c0d0acd3d.cql.sct-cluster-2.sct.scylladb.com because it's disabled in configuration
```

In other loaders happens the same.

Reply:
but new code relates only to scylla pods, loaders are done different way. We can try switching loaders to use the same resort as scylla pods, but I didn't plan it for this PR.

Reply:
I don't think we shouldn't touch the dynamic loader code for now, we can later align them to use the same. but it's not the target of this change.

Reply:
My bad, sorry. It is my loader solution lacks data because it uses chunk size >.<